lib/timeline/component/ItemSet.js

```diff
@@ -681,13 +681,37 @@ ItemSet.prototype.redraw = function() {
   // redraw the background group
   this.groups[BACKGROUND].redraw(range, nonFirstMargin, forceRestack);
 
-  // redraw all regular groups
-  util.forEach(this.groups, function (group) {
-    var groupMargin = (group == firstGroup) ? firstMargin : nonFirstMargin;
-    var groupResized = group.redraw(range, groupMargin, forceRestack);
-    resized = groupResized || resized;
-    height += group.height;
+  var redrawQueue = {};
+  var redrawQueueLength = 0;
+
+  // collect redraw functions
+  util.forEach(this.groups, function (group, key) {
+    if (key === BACKGROUND) return;
+    var groupMargin = group == firstGroup ? firstMargin : nonFirstMargin;
+    var returnQueue = true;
+    redrawQueue[key] = group.redraw(range, groupMargin, forceRestack, returnQueue);
+    redrawQueueLength = redrawQueue[key].length;
   });
+
+  if (redrawQueueLength) {
+    var redrawResults = {};
+
+    for (var i = 0; i < redrawQueueLength; i++) {
+      util.forEach(redrawQueue, function (fns, key) {
```

Comment:
Oof this stuff is almost over the top of my head.

But shouldn't this be: `util.forEach(redrawQueue[i]...`?

`redrawQueue` is a hash, of which every value is a queue of functions to run. I'm inferring that you want to run all function in the queue for every key of `redrawQueue`.

If I got it wrong here, please explain.

Reply:
Im sorry for bad naming :(

`group.redraw(range, groupMargin, forceRestack, returnQueue)` returns array of functions.

So redrawQueue is an queue, and every element in this queue is array of functions.

And when we release this queue, we call all functions from every element of queue.

Reply:
Nice to see you react to a question on the code. I was expecting @yotamberk, but I understand the original comes from you.

Will try to wrap my brain around it again. Perhaps some commenting would be in order here.